official/mnist/mnist.py

```diff
@@ -153,55 +165,122 @@ def train_input_fn():
         'image': image,
     })
-    mnist_classifier.export_savedmodel(FLAGS.export_dir, input_fn)
+    mnist_classifier.export_savedmodel(flags.export_dir, input_fn)
 
 
-if __name__ == '__main__':
-  parser = argparse.ArgumentParser()
-  parser.add_argument(
-      '--batch_size',
-      type=int,
-      default=100,
-      help='Number of images to process in a batch')
-  parser.add_argument(
-      '--data_dir',
-      type=str,
-      default='/tmp/mnist_data',
-      help='Path to directory containing the MNIST dataset')
-  parser.add_argument(
-      '--model_dir',
-      type=str,
-      default='/tmp/mnist_model',
-      help='The directory where the model will be stored.')
-  parser.add_argument(
-      '--train_epochs', type=int, default=40, help='Number of epochs to train.')
-  parser.add_argument(
-      '--data_format',
-      type=str,
-      default=None,
-      choices=['channels_first', 'channels_last'],
-      help='A flag to override the data format used in the model. channels_first '
-      'provides a performance boost on GPU but is not always compatible '
-      'with CPU. If left unspecified, the data format will be chosen '
-      'automatically based on whether TensorFlow was built for CPU or GPU.')
-  parser.add_argument(
-      '--export_dir',
-      type=str,
-      help='The directory where the exported SavedModel will be stored.')
+################################################################################
+# Multi-GPU helpers
+################################################################################
+def model_fn_multi(features, labels, mode, params):
+  """Wrapper for the model_fn that sets the optimizer for multi-GPU.
+
+  There are two key steps: wrap the optimizer, then replicate the model
+  function.
+  """
+  return model_fn(features, labels, mode, params, get_optimizer_multi)
+
+
+def get_optimizer_multi(learning_rate):
+  optimizer = tf.train.AdamOptimizer(learning_rate)
+  return tf.contrib.estimator.TowerOptimizer(optimizer)
+
+
+def get_batch_size_multi(current_size):
+  """For multi-gpu, batch-size must be a multiple of the number of
+  available GPUs.
+
+  TODO(karmel): This should eventually be handled by replicate_model_fn
+  directly. For now, doing the work here.
+  """
+  remainder = current_size % _get_num_gpu()
+  return current_size - remainder
+
 
+def _get_num_gpu():
+  """TODO(karmel): This should eventually be handled by replicate_model_fn
+  directly. For now, doing the work here.
+  """
+  local_device_protos = device_lib.list_local_devices()
+  return sum([1 for d in local_device_protos if d.device_type == 'GPU'])
+
+
+################################################################################
+# Run this script
+################################################################################
+def main(unused_argv):
+  model_function = model_fn_with_optimizer
+
+  if FLAGS.multi_gpu:
+    # Adjust batch size
```

Comment:
Though: @isaprykin - is that true? _must_ batch size be evenly divisible by the number of GPUs?

Reply:
As of now, yes, unfortunately.  replicate_model_fn could throw an exception if it's not the case, or devise a smarter scheme to do a partial run with only some towers...

Reply:
That one is a philosophical debate, and I come from the one-function-does-one-thing school of Python ;) More seriously, though, these are going to have to be used for Resnet, etc. as well, so this is a staging ground for extracting and sharing for all, rather than having that one liner spread across multiple files.

Reply:
Hmm...I don't quite follow the value of extracting something like:

```
x = (x // n) * n
``

into a library of shared functions.

In a bunch of these comments you're talking about resnet - so I think there is some mismatch between what the two of us think will be shared between this and resnet. Sorry about the confusion, but let's talk about this offline again.